File: publish_ghpages.py
#!/usr/bin/env python

from __future__ import absolute_import
from __future__ import print_function
import argparse
import os
import re
import shutil
import subprocess

ORIGINAL_REF = None


def shutdown(exit_code=0):
    if ORIGINAL_REF:
        subprocess.call(["git", "checkout", "--quiet", ORIGINAL_REF])
    exit(exit_code)


def run_command(args, exit_message=None):
    proc = subprocess.Popen(args, stdout=subprocess.PIPE)
    output, _ = proc.communicate()
    output = output.decode()
    exit_code = proc.returncode
    if exit_message and exit_code != 0:
        print("{}{}, aborting...".format(output, exit_message))
        shutdown(exit_code)
    return proc.returncode, output.strip()


def run_git_command(args, exit_message=None):
    return run_command(["git"] + args, exit_message)


_, ORIGINAL_COMMIT_SHA = run_git_command(["rev-parse", "HEAD"],
                                         "Can't get current git commit")
ORIGINAL_REF_RETURN_CODE, ORIGINAL_REF = run_git_command(["symbolic-ref",
                                                          "-q", "HEAD"])
if ORIGINAL_REF_RETURN_CODE == 0:
    ORIGINAL_REF = re.sub("^refs/heads/", "", ORIGINAL_REF)
else:
    ORIGINAL_REF = ORIGINAL_COMMIT_SHA


def staged_changes_in_repo(exit_message=None):
    exit_code, _ = run_git_command(["diff-index", "--exit-code", "--cached",
                                    "HEAD", "--"], exit_message)
    return exit_code


def main():
    parser = argparse.ArgumentParser(description="publish ghpages")
    parser.add_argument("remote", nargs="?", default="origin",
                        help="github remote name, default is 'origin'")
    args = parser.parse_args()

    remote = args.remote

    run_git_command(["status"], "Unable to determine git repo status")

    if not os.path.exists("setup.py"):
        print("Unable to locate setup.py file. This may not be a Python project.")
        shutdown(1)

    print("Validating that no git repo changes are outstanding...")
    staged_changes_in_repo("Staged changes in git repo")
    run_git_command(["diff-files", "--exit-code"],
                    "Unstaged changes in git repo")

    print("Running dist task...")
    run_command(["python", "dist.py"], "Dist task failed")

    print("Checking out gh-pages branch...")
    exit_code, _ = run_git_command(["rev-parse", "--quiet", "--verify",
                                    "gh-pages"])
    if exit_code == 0:
        run_git_command(["checkout", "gh-pages"],
                        "Unable to checkout gh-pages branch")
        run_git_command(["pull", remote, "gh-pages"])
    else:
        print("Local gh-pages branch does not exist so try to fetch it...")
        exit_code, _ = run_git_command(["fetch", remote, "gh-pages"])
        if exit_code == 0:
            run_git_command(["checkout", "-b", "gh-pages",
                             "remotes/{}/gh-pages".format(remote)])
        else:
            print("gh-pages does not exist in remote, so creating...")
            run_git_command(["checkout", "--orphan", "gh-pages"],
                            "Unable to create new gh-pages branch")
            run_git_command(["reset"], "Unable to reset local gh-pages branch")

    print("Copying latest generated docs to pydoc directory...")
    if os.path.exists("pydoc"):
        shutil.rmtree("pydoc")
    shutil.copytree("dist/doc", "pydoc")

    print("Ensuring .nojekyll file exists...")
    open(".nojekyll", "a").close()

    print("Adding changed files to git index...")
    run_git_command(["add", "pydoc/*"], "Unable to add pydoc/* files to index")
    run_git_command(["add", ".nojekyll"],
                    "Unable to add .nojekyll file to index")
    if staged_changes_in_repo() == 0:
        print("gh-pages branch already up to date, skipping commit...")
    else:
        print("Committing changes to gh-pages branch...")
        short_commit_sha = ORIGINAL_COMMIT_SHA[:7]
        run_git_command(["commit", "-m",
                         "Updating docs for {}".format(short_commit_sha)],
                        "Unable to commit changes")

    print("Pushing gh-pages changes to {}...".format(remote))
    run_git_command(["push", remote, "gh-pages"], "Unable to push changes")

    # Clean files so can switch back to the original ref. Preserve dist
    # directory so it's available for inspection, if desired.
    print("Cleaning up files before exit...")
    run_git_command(["clean", "-dfq", "-e", "dist"])

    print("Publish was successful, exiting...")
    shutdown()


main()

File: publish-ghpages.js
#!/usr/bin/env node

'use strict'

var childProcess = require('child_process')
var fs = require('fs')
var rimraf = require('rimraf')

var GH_PAGES_BRANCH = 'gh-pages'
var GH_PAGES_DIR = 'out'

var originalDir = process.cwd()

function shutdown (exitCode) {
  exitCode = exitCode || 0
  process.chdir(originalDir)
  process.exit(exitCode)
}

function runCommand (prog, args, exitMessage) {
  var proc = childProcess.spawnSync(prog, args)
  var output = proc.stdout.toString().trim()
  var exitCode = proc.status
  if (exitMessage && exitCode) {
    if (output) {
      console.log(output)
    }
    var err = proc.stderr.toString().trim()
    if (err) {
      console.log(err)
    }
    console.log(exitMessage + ', aborting...')
    shutdown(exitCode)
  }
  return {
    exitCode: proc.status,
    output: output
  }
}

function runGitCommand (args, exitMessage) {
  return runCommand('git', args, exitMessage)
}

function stagedChangesInRepo (exitMessage) {
  return runGitCommand(['diff-index', '--exit-code', '--cached', 'HEAD', '--'],
    exitMessage)
}

function main () {
  var remote = process.argv.length > 2 ? process.argv[2] : 'origin'

  runGitCommand(['status'], 'Unable to determine git repo status')

  if (!fs.existsSync('package.json')) {
    console.log(
      'Unable to locate package.json file. This may not be a Node.js project.')
    shutdown(1)
  }

  var originalCommitSha = runGitCommand(['rev-parse', 'HEAD'],
    "Can't get current git commit").output

  var refResult = runGitCommand(['symbolic-ref', '-q', 'HEAD'])
  var originalRef = refResult.exitCode ? originalCommitSha : refResult.output.replace('refs/heads/', '')

  console.log('Validating that no git repo changes are outstanding...')
  stagedChangesInRepo('Staged changes in git repo')
  runGitCommand(['diff-files', '--exit-code'], 'Unstaged changed in git repo')

  var ghPagesBranchUpToDate = false

  console.log('Setting up ' + GH_PAGES_BRANCH + ' branch...')
  if (runGitCommand(['rev-parse', '--quiet',
    '--verify', GH_PAGES_BRANCH]).exitCode) {
    console.log('Local ' + GH_PAGES_BRANCH +
      ' branch does not exist so try to fetch it')
    if (runGitCommand(['fetch', remote, GH_PAGES_BRANCH]).exitCode) {
      console.log(GH_PAGES_BRANCH + ' does not exist in remote, so creating...')
      runGitCommand(['checkout', '--orphan', GH_PAGES_BRANCH],
        'Unable to create new ' + GH_PAGES_BRANCH + ' branch')
      runGitCommand(['reset'], 'Unable to reset local ' + GH_PAGES_BRANCH +
        ' branch')
      runGitCommand(['commit', '--allow-empty', '-m', 'Initial commit'],
        'Unable to create initial ' + GH_PAGES_BRANCH + ' commit')
      runGitCommand(['clean', '-dfq'],
        'Unable to clean files before switching back to original branch')
      runGitCommand(['checkout', '--quiet', originalRef],
        'Unable to switch from ' + GH_PAGES_BRANCH +
        ' back to the original branch')
    } else {
      console.log('Creating local ' + GH_PAGES_BRANCH +
        ' branch tracking remote...')
      runGitCommand(['branch', GH_PAGES_BRANCH,
        'remotes/' + remote + '/' + GH_PAGES_BRANCH],
        'Unable to create local ' + GH_PAGES_BRANCH + ' branch')
    }
    ghPagesBranchUpToDate = true
  }

  console.log('Setting up worktree for ' + GH_PAGES_DIR + ' directory...')
  runGitCommand(['worktree', 'prune'], 'Unable to prune worktrees')
  var worktreeList = runGitCommand(['worktree', 'list']).output
  if (!worktreeList.match(process.cwd() + '/' +
      GH_PAGES_DIR + '.*\\[' + GH_PAGES_BRANCH + '\\]')) {
    console.log(GH_PAGES_BRANCH + ' worktree not setup, so creating...')
    rimraf.sync(GH_PAGES_DIR)
    runGitCommand(['worktree', 'add', GH_PAGES_DIR, GH_PAGES_BRANCH])
  }

  process.chdir(GH_PAGES_DIR)

  if (!ghPagesBranchUpToDate) {
    console.log('Pulling latest changes from remote into ' + GH_PAGES_BRANCH +
      ' branch')
    runGitCommand(['clean', '-dfq'],
      'Unable to clean files on ' + GH_PAGES_BRANCH + ' branch')
    runGitCommand(['pull', remote, GH_PAGES_BRANCH])
  }

  console.log('Running doc task...')
  runCommand('npm', ['run', 'doc'], 'Doc task failed')

  console.log('Adding changed files to git index...')
  runGitCommand(['add', 'jsdoc/*'], 'Unable to add jsdoc/* files to index')
  if (stagedChangesInRepo().exitCode) {
    console.log('Committing changes to ' + GH_PAGES_BRANCH + ' branch...')
    var shortCommitSha = originalCommitSha.substring(0, 7)
    runGitCommand(['commit', '-m', 'Updating docs for ' + shortCommitSha],
      'Unable to commit changes')
  } else {
    console.log(GH_PAGES_BRANCH +
      ' branch already up to date, skipping commit...')
  }

  console.log('Pushing ' + GH_PAGES_BRANCH + ' changes to ' + remote + '...')
  runGitCommand(['push', remote, GH_PAGES_BRANCH], 'Unable to push changes')

  console.log('Publish was successful, exiting...')
  shutdown()
}

main()
